Sources/SwiftKafka/KafkaProducer.swift

```diff
@@ -228,68 +230,161 @@ public actor KafkaProducer {
+        mutating func nextPollLoopAction() -> PollLoopAction {
+            switch self.state {
+            case .uninitialized:
+                fatalError("\(#function) invoked while still in state \(self.state)")
+            case .started(let client, _, _, _):
+                return .poll(client: client)
+            case .flushing(let client, let source):
+                if client.outgoingQueueSize > 0 {
+                    return .poll(client: client)
+                } else {
+                    self.state = .finished
+                    return .terminatePollLoopAndFinishSource(source: source)
+                }
+            case .finished:
+                return .terminatePollLoop
+            }
+        }
 
-    /// Check `topicHandles` for a handle matching the topic name and create a new handle if needed.
-    /// - Parameter topic: The name of the topic that is addressed.
-    private func createTopicHandleIfNeeded(topic: String) throws -> OpaquePointer? {
-        if let handle = self.topicHandles[topic] {
-            return handle
-        } else {
-            let newHandle = try self.client.withKafkaHandlePointer { handle in
-                let rdTopicConf = try RDKafkaTopicConfig.createFrom(topicConfig: self.topicConfig)
-                return rd_kafka_topic_new(
-                    handle,
-                    topic,
-                    rdTopicConf
+        /// Action to be taken when wanting to send a message.
+        enum SendAction {
+            /// Send the message.
+            ///
+            /// - Important: `newMessageID` is the new message ID assigned to the message to be sent.
+            case send(
+                client: KafkaClient,
+                newMessageID: UInt,
+                topicHandles: RDKafkaTopicHandles
+            )
+        }
+
+        /// Get action to be taken when wanting to send a message.
+        ///
+        /// - Returns: The action to be taken.
+        mutating func send() throws -> SendAction {
+            switch self.state {
+            case .uninitialized:
+                fatalError("\(#function) invoked while still in state \(self.state)")
+            case .started(let client, let messageIDCounter, let source, let topicHandles):
+                let newMessageID = messageIDCounter + 1
+                self.state = .started(
+                    client: client,
+                    messageIDCounter: newMessageID,
+                    source: source,
+                    topicHandles: topicHandles
                 )
-                // rd_kafka_topic_new deallocates topic config object
+                return .send(
+                    client: client,
+                    newMessageID: newMessageID,
+                    topicHandles: topicHandles
+                )
+            case .flushing, .finished:
+                throw KafkaError.connectionClosed(reason: "Tried to produce a message with a closed producer")
```

Comment:
Can we create a new separate error here when we flushing which indicates that we are currently shutting down and flushing please